core/drivers/imx_ocotp.c

```diff
@@ -140,6 +173,116 @@ TEE_Result imx_ocotp_read(unsigned int bank, unsigned int word, uint32_t *val)
 	return ret;
 }
 
+TEE_Result imx_ocotp_write(unsigned int bank, unsigned int word, uint32_t val)
+{
+	TEE_Result ret = TEE_ERROR_GENERIC;
+
+	if (!val)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (bank > g_ocotp->nb_banks || word > g_ocotp->nb_words)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (!g_ocotp->write_fuse)
+		return TEE_ERROR_NOT_IMPLEMENTED;
+
+	assert(g_base_addr && g_ocotp);
+
+	mutex_lock(&fuse_read);
+
+	ocotp_clock_enable();
+
+	/* Clear error bit */
+	io_write32(g_base_addr + OCOTP_CTRL_CLR, OCOTP_CTRL_ERROR);
+
+	/* Wait for busy flag to be cleared */
+	ret = ocotp_ctrl_wait_for(OCOTP_CTRL_BUSY);
+	if (ret) {
+		EMSG("OCOTP is busy");
+		goto out;
+	}
+
+	ret = g_ocotp->write_fuse(bank, word, val);
+
+	io_write32(g_base_addr + OCOTP_CTRL_SET, OCOTP_CTRL_RELOAD_SHADOWS);
+	udelay(1);
+
+	ret = ocotp_ctrl_wait_for(OCOTP_CTRL_BUSY);
+	if (ret) {
+		EMSG("OCOTP is busy");
+		goto out;
+	}
+
+	DMSG("OCOTP Bank %d Word %d Fuse 0x%" PRIx32, bank, word, val);
+out:
+	mutex_unlock(&fuse_read);
+
+	return ret;
+}
+
+static TEE_Result ocotp_mx8m_set_timing(void)
+{
+	uint32_t relax, strobe_read, strobe_prog;
+	uint32_t clk_rate;
+	uint32_t timing;
+
+	/* Assume the IPG_ROOT clock is running at 66.67 MHz */
+	clk_rate = 66666667;
+
+	relax = DIV_ROUND_UP(clk_rate * TIMING_RELAX_NS, 1000000000) - 1;
+
+	strobe_read = DIV_ROUND_UP(clk_rate * TIMING_STROBE_READ_NS, 1000000000);
+	strobe_read += 2 * (relax + 1) - 1;
+	strobe_prog = UDIV_ROUND_NEAREST(clk_rate * TIMING_STROBE_PROG_US, 1000000);
+	strobe_prog += 2 * (relax + 1) - 1;
+
+	timing = io_read32(g_base_addr + OCOTP_TIMING) & OCOTP_TIMING_WAIT;
+	timing |= field_prep(OCOTP_TIMING_RELAX, relax);
+	timing |= field_prep(OCOTP_TIMING_STROBE_READ, strobe_read);
+	timing |= field_prep(OCOTP_TIMING_STROBE_PROG, strobe_prog);
+
+	io_write32(g_base_addr + OCOTP_TIMING, timing);
+
+	return ocotp_ctrl_wait_for(OCOTP_CTRL_BUSY);
+}
+
+static TEE_Result ocotp_mx8m_write_fuse(unsigned int bank, unsigned int word, uint32_t val)
+{
+	TEE_Result ret;
+	uint32_t reg;
+
+	ret = ocotp_mx8m_set_timing();
+	if (ret) {
+		EMSG("OCOTP set_timing failed");
+		return ret;
+	}
+
+	/* Control register */
+	reg = io_read32(g_base_addr + OCOTP_CTRL);
+	reg &= ~OCOTP_CTRL_ADDR;
+	reg |= field_prep(OCOTP_CTRL_ADDR, OCOTP_ADDR(bank, word));
+	reg |= field_prep(OCOTP_CTRL_WR_UNLOCK, OCOTP_CTRL_WR_UNLOCK_KEY);
+	io_write32(g_base_addr + OCOTP_CTRL, reg);
+
+	/* Clear error bit */
+	io_write32(g_base_addr + OCOTP_CTRL_CLR, OCOTP_CTRL_ERROR);
+
+	io_write32(g_base_addr + OCOTP_DATA, val);
+	ret = ocotp_ctrl_wait_for(OCOTP_CTRL_BUSY);
+	if (ret)
+		return ret;
+
+	/* Write postamble */
+	udelay(2000);
```

Comment:
![image](https://github.com/user-attachments/assets/cb662e1b-73d7-42c2-8f5e-c708b36f9f87)
According to above image, We will be writing maximum of 32bits in one go and In that case, I think 400us is enough here.